### PR TITLE
[cephfs] add config doc for exporting CephFS over NFS

### PR DESCRIPTION
**Pull request by ajarr**
... using NFS-Ganesha NFS server.

Signed-off-by: Ramana raja <rraja@redhat.com>

### CODE REVIEW

**Comment by ajarr** [comment]
@jtlayton , I think I should add the config setting to turn off Ganesha caching, and turn on delegations. Should we also add the info about rados storing config and client recovery data as you did here,
https://review.gerrithub.io/#/c/403825/6/src/config_samples/ceph.conf
?

**Comment by mattbenjamin** [comment]
@ajarr that seems appropriate and useful, to me at least

File: doc/cephfs/nfs.rst [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+`NFS-Ganesha NFS server <https://github.com/nfs-ganesha/nfs-ganesha/wiki>`_.
+
+Requirements
+============
+
+-  Ceph filesystem (preferably latest luminous or later versions)
+-  'nfs-ganesha' and 'nfs-ganesha-ceph' packages (preferably the latest
+   ganesha v2.5 stable or later stable versions)
+
+   +  RPMs and packages for xenial available at http://download.ceph.com/nfs-ganesha/
```
[the remark this answers is not in the document]
Comment (ajarr) [reply doc/cephfs/nfs.rst]: Good point

**Comment by ajarr** [comment]
> I worry a little about this information "bitrotting" to some degree. We're tracking documentation within the ceph repo about an application that expect to be layered over it that is being developed separately. I'm not sure this is the right place for it.

@jtlayton, interesting. So you're thinking about the FSAL_CEPH/NFS-Ganesha developers having to update documentation in two places, NFS-Ganesha repo and Ceph repo?  

I think we at least need to have a placeholder in the Ceph docs that points to the appropriate documentation in the Ganesha repo to highlight the NFS-Ganesha over CephFS work.

Maybe I should just get rid of the text under the `NFS-Ganesha configuration` section in this doc, and add a link that points to  `config_samples/ceph.conf` (that you just updated) in the ganesha repo? This way I'm not duplicating documentation, and the NFS-Ganesha/FSAL developers don't need to update docs in both the repos in the future.

What do you think?

**Comment by batrick** [comment]
jenkins render docs

File: doc/cephfs/nfs.rst [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    CEPH
+    {
+	    # Ceph_Conf = /etc/ceph/ceph.conf;
+
+	    # umask = 0;
+    }
+
+EXPORT {} options
+
+- ``Export_ID`` must be a integer value.
```
Comment (batrick) [inline doc/cephfs/nfs.rst]: @jtlayton can you elaborate on the significance of the export ID? Does it need to be unique across a ganesha cluster?

File: doc/cephfs/nfs.rst [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+- ``Path`` is the CephFS dir path to export.
+
+- ``Pseudo`` is the pseudo FS path for NFSv4 exports.
```
Comment (batrick) [inline doc/cephfs/nfs.rst]: + Clients will use this path when mounting Ganesha.

File: doc/cephfs/nfs.rst [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+
+NFS-Ganesha configuration
+-------------------------
+
+A minimal ganesha.conf for configuring NFS-Ganesha over CephFS includes a
+EXPORT block and a CEPH block. Following is an example::
+
+    EXPORT
+    {
```
Comment (batrick) [inline doc/cephfs/nfs.rst]: I feel like this minimal configuration should also disable the Ganesha caching. Ganesha caching should be disabled 100% of the time and we should document taht.

File: doc/cephfs/nfs.rst [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+It is preferred to mount the NFS-Ganesha exports using NFSv4.1+ protocols
+to get the benefit of sessions.
+
+Conventions for mounting NFS resources are platform-specific. The
+following conventions work on Linux and some Unix platforms:
+
+From the command line::
+
+  mount -t nfs -o nfsvers=4.1,proto=tcp <ganesha-host-name>:/ <mount-point>
```
Comment (batrick) [inline doc/cephfs/nfs.rst]: `<ganesha-host-name>:<ganesha pseudo path>`

File: doc/cephfs/nfs.rst [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+Current limitations
+===================
+
+- A Ganesha server daemon cannot export more than one Ceph Filesystem.
```
Comment (batrick) [inline doc/cephfs/nfs.rst]: O.o

really?

**Comment by batrick** [comment]
Documenting each config options is okay with me, even if the documentation is in danger of rot / getting out-of-sync with Ganesha docs.

This document should emphasize **why** we have suggested/required certain options though. e.g. why is necessary to use `Protocols = 4;`? (Not all configs need this treatment of course.)

We also need a section for the RADOS backed exports.

File: doc/cephfs/nfs.rst (hunk 4 above)
Comment (renhwztetecs) [reply doc/cephfs/nfs.rst]: when we tested, we found that if nfs-ganehsa has cache, it will cause client cache pressure. Using Attr_Expiration_Time=0 does not close nfs-ganesha's cache, Is there any other parameter to turn off the nfs-ganehsa parameter?

File: doc/cephfs/nfs.rst (hunk 6 above)
Comment (renhwztetecs) [reply doc/cephfs/nfs.rst]: Can export multiple directories, as follows:


>EXPORT
{
        Export_Id = 11;
        Path = /test1;
        Pseudo = /test1;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}
EXPORT
{
        Export_Id = 12;
        Path = /test2;
        Pseudo = /test2;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}

File: doc/cephfs/nfs.rst (hunk 4 above)
Comment (ajarr) [reply doc/cephfs/nfs.rst]: > I feel like this minimal configuration should also disable the Ganesha caching. Ganesha caching should be disabled 100% of the time and we should document taht.

@batrick understood.

Comment (ajarr) [reply doc/cephfs/nfs.rst]: @renhwztetecs , did you also try disabling or minimizing Ganesha's dentry, open fd, inode caching as mentioned here,
https://review.gerrithub.io/#/c/403825/6/src/config_samples/ceph.conf
Please see line 53 to 68

File: doc/cephfs/nfs.rst (hunk 5 above)
Comment (ajarr) [reply doc/cephfs/nfs.rst]: Good catch!

File: doc/cephfs/nfs.rst (hunk 6 above)
Comment (ajarr) [reply doc/cephfs/nfs.rst]: @renhwztetecs ,  yes a single Ganesha daemon can export multiple directories within a Ceph file system. Have you been able to export multiple file systems?

File: doc/cephfs/nfs.rst (hunk 3 above)
Comment (ajarr) [reply doc/cephfs/nfs.rst]: Will add it

**Comment by ajarr** [comment]
@batrick 

> Documenting each config options is okay with me, even if the documentation is in danger of rot / getting out-of-sync with Ganesha docs.

OK.

> This document should emphasize why we have suggested/required certain options though. e.g. why is necessary to use Protocols = 4;? (Not all configs need this treatment of course.)

We say that "It is preferred to mount the NFS-Ganesha exports using NFSv4.1+ protocols
to get the benefit of sessions"  that's not sufficient?
https://github.com/ceph/ceph/pull/20915/commits/67be0819acbd03d2bdb355a3be5acccad0a4d424#diff-6b6c587a347d8eab90ea6f18e3511bffR144

File: doc/cephfs/nfs.rst (hunk 6 above)
Comment (renhwztetecs) [reply doc/cephfs/nfs.rst]: @ajarr 
Yes, I can export both FSAL_VFS and FASL Ceph(Version: nfs-ganesha-2.5.4)
>EXPORT
{
        Export_Id = 14;
        Path = /test4;
        Pseudo = /test4;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}
EXPORT
{
        Export_Id = 77;
        Path = /abc;
        Pseudo = /abc;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = VFS;
        }
}

File: doc/cephfs/nfs.rst (hunk 4 above)
Comment (renhwztetecs) [reply doc/cephfs/nfs.rst]: @ajarr Thank you very much for your hints.

@ajarr @batrick @jtlayton 
In addition, I think that NFS-Ganehsa lacks a forget mechanism like FUSE(calling ceph's Client::ll_forget()) to periodically release the cache mechanism in libcephfs because of the large number of files passing through NFS-Ganehsa. When writing, the "failing to respond to cache pressure" warnings alert will appear on the libcephfs side. See this link: http://tracker.ceph.com/issues/12334
Any suggestions are very grateful?

**Comment by batrick** [comment]
> We say that "It is preferred to mount the NFS-Ganesha exports using NFSv4.1+ protocols to get the benefit of sessions" that's not sufficient?
67be081#diff-6b6c587a347d8eab90ea6f18e3511bffR144

That's a little vague but okay for now. WE can improve it later.

File: doc/cephfs/nfs.rst (hunk 6 above)
Comment (batrick) [reply doc/cephfs/nfs.rst]: Let's reword it to say "FSAL_CEPH can only export one Ceph file system although multiple directories in a Ceph file system may be exported."

File: doc/cephfs/nfs.rst (hunk 4 above)
Comment (batrick) [reply doc/cephfs/nfs.rst]: @renhwztetecs if you have an example reproducer, please post to that issue.

**Comment by jtlayton** [comment]
> Maybe I should just get rid of the text under the NFS-Ganesha configuration section in this doc, and add a link that points to config_samples/ceph.conf (that you just updated) in the ganesha repo? This way I'm not duplicating documentation, and the NFS-Ganesha/FSAL developers don't need to update docs in both the repos in the future.

I think this doc probably ought to be transplanted into the nfs-ganesha sources. The sample config file there is a starting point, but a more thorough writeup like this could also be added. A small doc inside ceph for this is fine, but really ought to just contain a minimum nfs-ganesha version required for functionality.

Given that the ganesha codebase doesn't live inside ceph, this will be a constant struggle to keep things in sync: What if they change ganesha's config options in the future? Your doc will have to have sections for old and new and there will be an inevitable lag in getting them updated. When they're part of the ganesha sources, then that should mostly get updated as people change the code.

File: doc/cephfs/nfs.rst (hunk 6 above)
Comment (ajarr) [reply doc/cephfs/nfs.rst]: Done

**Comment by ajarr** [comment]
> I think this doc probably ought to be transplanted into the nfs-ganesha sources. The sample config file there is a starting point, but a more thorough writeup like this could also be added. A small doc inside ceph for this is fine, but really ought to just contain a minimum nfs-ganesha version required for functionality.

I added more than just minimum nfs-ganesha version. I couldn't find an appropriate place in nfs-ganesha repo to put the details mentioned here. Didn't want to delay this patch further.

> Given that the ganesha codebase doesn't live inside ceph, this will be a constant struggle to keep things in sync: What if they change ganesha's config options in the future? Your doc will have to have sections for old and new and there will be an inevitable lag in getting them updated. When they're part of the ganesha sources, then that should mostly get updated as people change the code.

Added only setup and usage steps in Ceph repo that shouldn't get obsolete. I point to your doc for ganesha config options,
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/ceph.conf
Hope this works for now.

**Comment by ajarr** [comment]
jenkins render doc

**Comment by batrick** [comment]
jenkins render docs
